dhall/src/Dhall/DirectoryTree.hs

```diff
@@ -215,12 +222,39 @@ toDirectoryTree allowSeparators path expression = case expression of
         empty
 
     process key value = do
-        when (not allowSeparators && Text.isInfixOf (Text.pack [ FilePath.pathSeparator ]) key) $
-            die
-
-        Directory.createDirectoryIfMissing allowSeparators path
-
-        toDirectoryTree allowSeparators (path </> Text.unpack key) value
+        case keyPathSegments of
+            -- Fail if path is absolute, which is a security risk.
+            "" : _ ->
+                die
+            -- Detect Windows absolute paths like "C:".
+            [_ , ':'] : _ ->
+                die
+            -- Fail if separators are not allowed by the option.
+            _ : _ | not allowSeparators ->
+                die
+            _ ->
+                return ()
+
+        -- Fail if path contains attempts to go to container directory,
+        -- which is a security risk.
+        if elem ".." keyPathSegments
+            then die
+            else return ()
+
+        (dirPath, fileName) <- case reverse keyPathSegments of
+            h : t ->
+                return
+                    ( Foldable.foldl' (</>) path (reverse t)
+                    , h )
+            _ ->
+                die
+
+        Directory.createDirectoryIfMissing True dirPath
+
+        toDirectoryTree allowSeparators (dirPath </> fileName) value
```

Comment:
I think you want something like this:

```suggestion
        (dirPath', fileName) <- case reverse keyPathSegments of
            h : t ->
                return
                    ( Foldable.foldl' (</>) path (reverse t)
                    , h )
            _ ->
                die

        let dirPath = path </> dirPath'
        
        Directory.createDirectoryIfMissing True dirPath

        toDirectoryTree allowSeparators (dirPath </> fileName) value
```

In other words, `dirPath` needs to incorporate `path` as a prefix, otherwise the `Map` fields will be created as subdirectories relative to `PWD` when they should be relative to `path`.

Note that https://github.com/nikita-volkov/dhall-haskell/pull/1 fixes this, too